### PR TITLE
[DNM] Redrop Metadata Server Protocol and Fix Firewall/Certificates in Update

### PR DESCRIPTION
**Pull request by byue**
## Description

This PR redrops MDS support, ensures WireServer firewall rule is set before goal state is queried after agent update, ensures WireServer transport certificates are generated, cleans up MetadataServer transport certificates, and updates protocol file/WireServer endpoint file during update. Most of my fixes are in env.py and update.py.

---

### Do Not Merge Until the Following is Complete
- [ ] Final Validation (extension test matrix scenarios) with tip of develop + this PR
- [ ] Run extension tests in AzureStack team CI

### Quality of Code and Contribution Guidelines
- [x] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

---
This change is [Reviewable](https://reviewable.io/reviews/azure/walinuxagent/1803)

### CODE REVIEW

**Comment by narrieta** [comment]
@byue - Some unit tests are failing... could you take a look? (click on the Travis CI link above)

File: azurelinuxagent/ga/env.py [hunk 1]
```diff
@@ -80,6 +82,13 @@ def run(self):
         self.dhcp_handler.conf_routes()
         self.hostname = self.osutil.get_hostname_record()
         self.dhcp_id_list = self.get_dhcp_client_pid()
+        # Cleanup MDS firewall rule and ensure WireServer firewall rule is set
+        # before we query goal state (for agents migrating from MDS protocol WS rule is not set)
+        # We are setting firewall rules before thread is spun off to avoid race condition
+        # where we query goal state in update before setting firewall rules.
+        if conf.enable_firewall() and remove_metadata_server_firewall_rule:
+            self.osutil.remove_firewall(METADATA_SERVER_ENDPOINT, uid=os.getuid())
```
Comment (byue) [inline azurelinuxagent/ga/env.py]: always remove MDS firewall rule or is this check ok?

**Comment by byue** [comment]
> The changes on the update handler are related to protocol (maybe protocol detection). They should be somewhere in the protocol module.

Created a new file called migration_util to put in migratiion handling logic. Update.py is now cleaner and this migration_util file can be removed easily once everyone is on WireServer.

File: azurelinuxagent/ga/env.py (hunk 1 above)
Comment (byue) [reply azurelinuxagent/ga/env.py]: Discussed with Norberto, IMDS has same IP but is not protected in public so it is ok to remove 169.254.169.254 from firewall rules (Azure does not have 169.254.169.254 in firewall rules).

File: azurelinuxagent/common/protocol/util.py [hunk 2]
```diff
@@ -186,7 +158,7 @@ def get_wireserver_endpoint(self):
 
         return self.endpoint
 
-    def _set_wireserver_endpoint(self, endpoint):
+    def set_wireserver_endpoint(self, endpoint):
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/util.py]: Let's leave this one marked as private and let the migration utils reach into the private.
Idea is that this method should be directly called from code outside of protocol/*.
Same for save_protocol

File: azurelinuxagent/ga/env.py [hunk 3]
```diff
@@ -80,6 +81,13 @@ def run(self):
         self.dhcp_handler.conf_routes()
         self.hostname = self.osutil.get_hostname_record()
         self.dhcp_id_list = self.get_dhcp_client_pid()
+        # Cleanup MDS firewall rule and ensure WireServer firewall rule is set
```
Comment (narrieta) [inline azurelinuxagent/ga/env.py]: the env thread is meant to check for required  configuration (e.g. wireserver firewall rules) periodically.
isn't the migration supposed to be happen only 1 time?

File: azurelinuxagent/common/protocol/migration_util.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Microsoft Azure Linux Agent
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/migration_util.py]: we should add unit tests for this module

File: azurelinuxagent/common/protocol/migration_util.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+import os
+import azurelinuxagent.common.conf as conf
+
+from azurelinuxagent.common.protocol.goal_state import TRANSPORT_CERT_FILE_NAME
+from azurelinuxagent.common.protocol.util import WIRE_PROTOCOL_NAME
+from azurelinuxagent.common.utils.restutil import KNOWN_WIRESERVER_IP
+
+# MetadataServer Legacy Certificates for Cleanup
+LEGACY_TRANSPORT_PRV_FILE_NAME = "V2TransportPrivate.pem"
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/migration_util.py]: let's mark all those constants as private

Comment (narrieta) [inline azurelinuxagent/common/protocol/migration_util.py]: I would name those METADATA somewhere on on those names, and would probably rename the file to metadata_migration_util.

File: azurelinuxagent/common/protocol/migration_util.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+METADATA_SERVER_ENDPOINT = "169.254.169.254"
+
+def is_migrating_protocol():
+    """
+    Migrating away from Metadata Server protocol if Metadata Server transport
+    certificate is present or Wire Server transport certificate is missing.
+    """
+    transport_cert_file = os.path.join(conf.get_lib_dir(), TRANSPORT_CERT_FILE_NAME)
+    legacy_transport_cert_file = os.path.join(conf.get_lib_dir(), LEGACY_TRANSPORT_CERT_FILE_NAME)
+    return os.path.isfile(legacy_transport_cert_file) or not os.path.isfile(transport_cert_file)
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/migration_util.py]: if we check for the wireserver transport certificate then this code has a (timing) dependency on the code that creates the certificate. Isn't it enough to check for the metadata certificate?

**Pull request closed without merging**